### PR TITLE
Notion Bug Squashing

### PR DESCRIPTION
**Pull request by joshuatazrein**
- [ ]  [Notion import error "Folder already exists" #83](https://github.com/obsidianmd/obsidian-importer/issues/83)
    - It seems like the change should be incorporated by the Obsidian team in `main`. Let me know how that sounds.
- [x] [Equations are not imported correctly from Notion #81](https://github.com/obsidianmd/obsidian-importer/issues/81)
- [x] [Importer not correctly handling Notion Teamspace update #79](https://github.com/obsidianmd/obsidian-importer/issues/79)
    - I tried this with a Teamspace export and it worked perfectly. Let me know if there is an edge case I missed.
- [x] [Notion code blocks are imported without line breaks #82](https://github.com/obsidianmd/obsidian-importer/issues/82)
- [x] [Incorrect formatting of Notion bookmark #76](https://github.com/obsidianmd/obsidian-importer/issues/76)
- [x] [Notion import corrupted when page includes "Toggle Heading" #75](https://github.com/obsidianmd/obsidian-importer/issues/75)

### CODE REVIEW

File: src/formats/notion/convert-to-md.ts [hunk 1]
```diff
@@ -283,16 +315,51 @@ function fixNotionDates(body: HTMLElement) {
...
+					return
+				}
+			}
+		}
+		
+	})
+}
+
+function replaceElementsWithChildren(body: HTMLElement, selector: string) {
+	let element = body.querySelector(selector)
```
Comment (lishid) [inline src/formats/notion/convert-to-md.ts:344]: Why not use findAll?

File: src/formats/notion/convert-to-md.ts [hunk 2]
```diff
@@ -193,6 +200,28 @@ function fixDoubleBackslash(markdownBody: string) {
 	return markdownBody;
 }
 
+function fixEquations(body: HTMLElement) {
+	const katexEls = body.findAll('.katex')
+	katexEls.forEach(katex => {
```
Comment (lishid) [inline src/formats/notion/convert-to-md.ts:205]: Can we convert all of these forEach to for-of loops instead?

File: src/formats/notion/convert-to-md.ts [hunk 3]
```diff
@@ -193,6 +200,28 @@ function fixDoubleBackslash(markdownBody: string) {
...
+		const annotation = katex.querySelector('annotation')
```
Comment (lishid) [inline src/formats/notion/convert-to-md.ts:206]: Prefer to use `.find` over `.querySelector`.

**Review by lishid — COMMENTED** [review]
Can you run the auto formatter? There's some missing semicolons.

File: src/formats/notion/convert-to-md.ts (hunk 1 above)
Comment (joshuatazrein) [reply src/formats/notion/convert-to-md.ts:344]: Searching for the first element in a `while` loop ensured that the element hasn't been removed in a previous iteration (i.e. nested children etc). Was having trouble with the `findAll` function

Comment (lishid) [reply src/formats/notion/convert-to-md.ts]: Hmmm it seems that we are only removing the exact element being found while all nested children are reattached to the parent, so there shouldn't be any elements that gets removed from a previous iteration, no?

Comment (joshuatazrein) [reply src/formats/notion/convert-to-md.ts]: I can test using `findAll`.

File: src/formats/notion/convert-to-md.ts [hunk 4]
```diff
@@ -193,6 +200,28 @@ function fixDoubleBackslash(markdownBody: string) {
 	return markdownBody;
 }
 
+function fixEquations(body: HTMLElement) {
+	const katexEls = body.findAll('.katex');
+	for (let katex of katexEls) {
+		const annotation = katex.find('annotation');
+		if (!annotation) return;
```
Comment (lishid) [inline src/formats/notion/convert-to-md.ts:207]: This should be `continue`.

File: src/formats/notion/convert-to-md.ts [hunk 5]
```diff
@@ -283,16 +315,51 @@ function fixNotionDates(body: HTMLElement) {
...
+		}
+		
+	});
+}
+
+function replaceElementsWithChildren(body: HTMLElement, selector: string) {
+	let els = body.findAll(selector);
+	for (let el of els) {
+		const children = el.children;
+		const childNodes: Element[] = [];
```
Comment (lishid) [inline src/formats/notion/convert-to-md.ts:348]: Would `Array.from(el.children)` here simplify things?

**Review by lishid — APPROVED** [review]
There's a couple of forEach to change (and be careful of `return` statements) but otherwise this looks great!

**Pull request merged**